xds/src/test/java/io/grpc/xds/AbstractXdsClientTest.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.truth.Truth.assertThat;
+import static io.grpc.xds.XdsClientTestHelper.buildCluster;
+import static io.grpc.xds.XdsClientTestHelper.buildClusterLoadAssignment;
+import static io.grpc.xds.XdsClientTestHelper.buildDiscoveryResponse;
+import static io.grpc.xds.XdsClientTestHelper.buildDropOverload;
+import static io.grpc.xds.XdsClientTestHelper.buildLbEndpoint;
+import static io.grpc.xds.XdsClientTestHelper.buildListener;
+import static io.grpc.xds.XdsClientTestHelper.buildLocalityLbEndpoints;
+import static io.grpc.xds.XdsClientTestHelper.buildRouteConfiguration;
+import static io.grpc.xds.XdsClientTestHelper.buildVirtualHost;
+
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Any;
+import io.envoyproxy.envoy.config.core.v3.HealthStatus;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse;
+import io.grpc.xds.AbstractXdsClient.MessagePrinter;
+import io.grpc.xds.AbstractXdsClient.ResourceType;
+import java.util.List;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link AbstractXdsClient}.
+ */
+@RunWith(JUnit4.class)
+public class AbstractXdsClientTest {
```

Comment:
This has only `MessagePrinter` tests and `MessagePrinter` is only used in log messages and not for any functionality. So these tests don't test any functionality, is that correct?

Reply:
The XdsClient's behaviors (including functionalities such as ADS stream closed and retry implemented in the abstract laye) are tested in each of the concrete implementations (see `ClientXdsClientTest` and `ServerXdsClientTest`). In high-level, it's more like a blackbox test.